lib/Sema/CSDiagnostics.cpp

```diff
@@ -4629,19 +4629,67 @@ bool InaccessibleMemberFailure::diagnoseAsError() {
   auto loc = nameLoc.isValid() ? nameLoc.getStartLoc() : anchor->getLoc();
   auto accessLevel = Member->getFormalAccessScope().accessLevelForDiagnostics();
   if (auto *CD = dyn_cast<ConstructorDecl>(Member)) {
-    emitDiagnostic(loc, diag::init_candidate_inaccessible,
-                   CD->getResultInterfaceType(), accessLevel)
-        .highlight(nameLoc.getSourceRange());
+    diagnoseInaccessibleInitializer(CD, loc, nameLoc, accessLevel);
   } else {
     emitDiagnostic(loc, diag::candidate_inaccessible, Member->getBaseName(),
                    accessLevel)
         .highlight(nameLoc.getSourceRange());
+    emitDiagnostic(Member, diag::decl_declared_here, Member->getFullName());
   }
 
-  emitDiagnostic(Member, diag::decl_declared_here, Member->getFullName());
   return true;
 }
 
+void InaccessibleMemberFailure::diagnoseInaccessibleInitializer(
+    ConstructorDecl *CD, SourceLoc Loc, DeclNameLoc NameLoc,
+    AccessLevel AccessLevel) {
+  // If this is a memberwise or default initializer, say so and explain why it
+  // was synthesized with a certain access level.
+  auto ICK = CD->getImplicitConstructorKind();
+  if (ICK && (*ICK == ImplicitConstructorKind::Default ||
+              *ICK == ImplicitConstructorKind::Memberwise)) {
+    // If the initializer is in a different module, it must be made public to be
+    // accessible from the anchor expression.
+    bool requiresPublicAccess =
+        CD->getModuleContext() != getConstraintSystem().DC->getParentModule();
+    // Noting properties isn't helpful for default initializers or if the
+    // inititalizer needs to be made public to be usable in this context.
+    SmallVector<VarDecl *, 3> propertiesContributingToAccessLevel;
+    if (!requiresPublicAccess && *ICK == ImplicitConstructorKind::Memberwise) {
+      auto structDecl =
+          CD->getResultInterfaceType()->getStructOrBoundGenericStruct();
+      assert(structDecl &&
+             "only structs support default memberwise initializers");
+      for (auto member : structDecl->getMembers()) {
+        auto var = dyn_cast<VarDecl>(member);
+        if (!var)
+          continue;
+
+        // If a property participates in memberwise initialization and has the
+        // same access level as the memberwise initializer, it must have
+        // contributed to the initializer's access level.
+        if (var->isMemberwiseInitialized(/*preferDeclaredProperties=*/true) &&
+            var->getFormalAccess() == AccessLevel)
+          propertiesContributingToAccessLevel.push_back(var);
+      }
+    }
+    emitDiagnostic(Loc, diag::memberwise_init_candidate_inaccessible,
+                   static_cast<unsigned>(*ICK), CD->getResultInterfaceType(),
+                   AccessLevel, requiresPublicAccess)
+        .highlight(NameLoc.getSourceRange());
```

Comment:
I think you can just pass `NameLoc` directly.

Reply:
`NameLoc` is a `DeclNameLoc`, so there's no implicit conversion

Reply:
Ah, okay.